golang docs (#413)
ingonyama-zk/icicle: docs/docs/icicle/golang-bindings.md

# Golang bindings

Golang bindings allow you to use ICICLE as a golang library.
The source code for all Golang libraries can be found [here](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang).

The Golang bindings are comprised of multiple packages.

> [`core`](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang/core) which defines all common shared methods and structures, such as configuration structures, or memory slices.

Reviewer: ```suggestion
[`core`](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang/core) which defines all shared methods and structures, such as configuration structures, or memory slices.
```

> [`cuda-runtime`](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang/cuda_runtime) which defines abstractions for CUDA methods for allocating memory, initializing and managing streams. As well as `DeviceContext` which enables users to define and keep track of devices.

Reviewer: ```suggestion
[`cuda-runtime`](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang/cuda_runtime) which defines abstractions for CUDA methods for allocating memory, initializing and managing streams, and `DeviceContext` which enables users to define and keep track of devices.
```

> Each curve has its own package, you can find all curves [here](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang/curves). If you project uses BN-254 you only need to install that single package named [`bn254`](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang/curves/bn254).

Reviewer: ```suggestion
Each curve has its own package which you can find [here](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang/curves). If your project uses BN254 you only need to install that single package named [`bn254`](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang/curves/bn254).
```

> Each curve has its own package, you can find all curves [here](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang/curves). If you project uses BN-254 you only need to install that single package named [`bn254`](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang/curves/bn254).

Reviewer: >If you project uses BN-254 you only need to install that single package named [`bn254`](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang/curves/bn254).

What do you mean by install? The mod file will require `github.com/ingonyama-zk/icicle` regardless of what subpackage is used in `go get`

## Using ICICLE Golang bindings in your project

> Too add ICICLE to your `go.mod` file.

Reviewer: ```suggestion
To add ICICLE to your `go.mod` file.
```

```bash

> go get github.com/ingonyama-zk/icicle/goicicle

Reviewer: ```suggestion
go get github.com/ingonyama-zk/icicle
```

```

If you want to specify a specific branch

```bash

> go get github.com/ingonyama-zk/icicle/goicicle@<branch_name>

Reviewer: ```suggestion
go get github.com/ingonyama-zk/icicle@<branch_name>
```

```

For a specific commit

```bash

> go get github.com/ingonyama-zk/icicle/goicicle@<commit_id>

Reviewer: ```suggestion
go get github.com/ingonyama-zk/icicle@<commit_id>
```

```

To build the shared libraries you can run this script:

```
./build <curve> [G2_enabled]

curve - The name of the curve to build or "all" to build all curves
G2_enabled - Optional - To build with G2 enabled 
```

For example if you want to build all curves with G2 enabled you would run:

```bash
./build.sh all ON
```

If you are interested in building a specific curve you would run:

```bash

> ./build.sh bls12381 ON

Reviewer: ```suggestion
./build.sh bls12_381 ON
```

```

After building your shared libraries. You must export them so your system will be aware of their existence.

```bash
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH/<path_to_shared_libs>

> ```

Reviewer: I don't think this is necessary anymore since we are using static libraries
